crates/network/src/stream_pull.rs

```diff
@@ -6,26 +6,69 @@
 
 // TODO: Decide whether to model this as an abstract stream interface with the protocol identifier as an argument or
 
-use futures_util::{AsyncReadExt, AsyncWriteExt, Stream};
+use std::io;
+
+use futures_util::{Stream, StreamExt};
 use libp2p::{PeerId, StreamProtocol};
 use libp2p_stream::{AlreadyRegistered, Control, OpenStreamError};
 use serde::{Serialize, de::DeserializeOwned};
-use tokio::io::{AsyncRead, AsyncWrite};
-use tokio_stream::StreamExt;
+use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
 use tokio_util::compat::FuturesAsyncReadCompatExt;
 
+use crate::utils::{FixedAsyncRead, FixedAsyncWrite};
+
 /// The protocol identifier for Hypha's tensor streaming protocol.
 ///
 /// This constant defines the libp2p protocol string used for tensor data streaming
 /// between peers. It follows the libp2p convention of using a path-like identifier.
 const TENSOR_STREAM_PROTOCOL: StreamProtocol = StreamProtocol::new("/hypha-tensor-stream/pull");
 
-/// The maximum size of the resource header in bytes.
+/// The maximum size of the request header in bytes.
 ///
-/// This constant defines the maximum size of the resource header that can be
+/// This constant defines the maximum size of the request header that can be
 /// deserialized from the stream. It is used to prevent excessive memory usage
 /// and potential denial-of-service attacks.
-const MAX_RESOURCE_HEADER_SIZE: u64 = 1024 * 1024;
+const MAX_REQUEST_HEADER_SIZE: u64 = 1024 * 1024;
+
+/// The fixed header length used for announcing payload size.
+const PAYLOAD_LENGTH_HEADER_SIZE: usize = 8;
```

Comment:
Nit: Can be a bit clearer with

```suggestion
const PAYLOAD_LENGTH_HEADER_SIZE: usize = size_of::<u64>();
```